File: LeetCode/Java/src/problems/p0540_SingleElementInSortedArray/P0540.java
package problems.p0540_SingleElementInSortedArray;

public class P0540 {
    public int singleNonDuplicate(int[] nums) {
        /*
         * Solution v1: XOR
         * this is not a O(log n) solution
         *
         * Runtime: 0 ms @ (beats) 100%
         * Memory Usage: 39.2 MB @ (beats) 61%
         */
        // int xor = 0;
        // for (int num : nums) {
        //     xor ^= num;
        // }
        //
        // return xor;

        /*
         * (learn from Internet)
         * Solution v2: Binary Search
         */
        int left = 0, right = nums.length - 1;

        while (left < right) {
            // We want the first element of the middle pair,
            // which should be at an even index if the left part is sorted.
            // Example:
            // Index: 0 1 2 3 4 5 6
            // Array: 1 1 3 3 4 8 8
            //            ^
            int mid = left + (right - left) / 2;
            if (mid % 2 == 1) {
                mid--;
            }

            // We didn't find a pair. The single element must be on the left.
            // (pipes mean left & right)
            // Example: |0 1 1 3 3 6 6|
            //               ^ ^
            // Next:    |0 1 1|3 3 6 6
            if (nums[mid] != nums[mid + 1]) {
                right = mid;
            }

            // We found a pair. The single element must be on the right.
            // Example: |1 1 3 3 5 6 6|
            //               ^ ^
            // Next:     1 1 3 3|5 6 6|
            else {
                left = mid + 2;
            }
        }

        // 'start' should always be at the beginning of a pair.
        // When 'left > right', start must be the single element.
        return nums[left];
    }
}

File: LeetCode/Python/solutions/p0540.py
from typing import List


class Solution:
    def singleNonDuplicate(self, nums: List[int]) -> int:
        if len(nums) == 1:
            return nums[0]

        left = 0
        right = len(nums) - 1

        while left < right:
            mid = (left + right) // 2

            if mid % 2 != 0:
                mid -= 1

            if nums[mid] != nums[mid + 1]:
                right = mid
            else:
                left = mid + 2

        return nums[left]


if __name__ == "__main__":
    solution = Solution()

    nums = [1, 1, 2, 3, 3, 4, 4, 8, 8]
    print(f"return: {solution.singleNonDuplicate(nums)}")  # 2

    nums = [3, 3, 7, 7, 10, 11, 11]
    print(f"return: {solution.singleNonDuplicate(nums)}")  # 10
